metadata/metadata.go

```diff
@@ -72,42 +94,55 @@ func (hi HeaderInfo) Validate() error {
 	return nil
 }
 
+func (hi *HeaderInfo) Write(p []byte) (n int, err error) {
+	if err := decode(p, hi); err != nil {
+		return 0, err
+	}
+	return len(p), nil
+}
+
 // TypeInfo provides information of type of individual updates
 // archived in artifacts archive.
 type TypeInfo struct {
-	Rootfs string `json:"rootfs"`
+	Type string `json:"type"`
 }
 
 // Validate validates corectness of TypeInfo.
 func (ti TypeInfo) Validate() error {
-	if len(ti.Rootfs) == 0 {
+	if len(ti.Type) == 0 {
 		return ErrValidatingData
 	}
 	return nil
 }
 
+func (ti *TypeInfo) Write(p []byte) (n int, err error) {
+	if err := decode(p, ti); err != nil {
+		return 0, err
+	}
+	return len(p), nil
+}
+
 // Metadata contains artifacts metadata information. The exact metadata fields
 // are user-defined and are not specified. The only requirement is that those
 // must be stored in a for of JSON.
 // The only fields which must exist are 'DeviceType' and 'ImageId'.
-type Metadata struct {
-	// we don't know exactly what type of data we will have here
-	data map[string]interface{}
-}
+type Metadata map[string]interface{}
 
 // Validate check corecness of artifacts metadata. Since the exact format is
 // nost specified we are only checking if those could be converted to JSON.
 // The only fields which must exist are 'DeviceType' and 'ImageId'.
 func (m Metadata) Validate() error {
-	if m.data == nil {
+	log.Debugf("validating metadata: %v", m)
+	if m == nil {
 		return ErrValidatingData
 	}
 	// mandatory fields
 	var deviceType interface{}
 	var imageID interface{}
 
-	for k, v := range m.data {
+	for k, v := range m {
 		if v == nil {
+			log.Error("empty element while validating metedata")
 			return ErrValidatingData
 		}
 		if strings.Compare(k, "DeviceType") == 0 {
```

Comment:
if this is a JSON key, probably would be more natural to lean towards lower case snake style